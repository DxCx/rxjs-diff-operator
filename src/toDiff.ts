import { Observable } from 'rxjs/Observable';
import { Operator } from 'rxjs/Operator';
import { Subscriber } from 'rxjs/Subscriber';
import { IObservableDiff } from './interfaces';
import { diff } from 'deep-diff';

export function toDiff<T>(): Observable<IObservableDiff> {
  return this.lift(new ToDiffOperator());
}

export class ToDiffOperator<T> implements Operator<T, T> {
  public call(subscriber: Subscriber<T>, source: any): any {
    return source._subscribe(new ToDiffSubscriber(subscriber));
  }
}

class ToDiffSubscriber<T> extends Subscriber<T> {
  private count: number = 0;
  private isObject: boolean = false;
  private lastValue: any;

  constructor(destination: Subscriber<T>) {
    super(destination);
  }

  protected _next(value: T) {
      if ( 0 === this.count ) {
          this.isObject = (typeof value === 'object');
          this.destination.next({type: 'init', payload: value, isObject: this.isObject});
      } else if ( this.isObject ) {
          this.destination.next({type: 'update', payload: diff(this.lastValue, value)});
      } else {
          this.destination.next({type: 'update', payload: value});
      }

      this.lastValue = value;
      this.count ++;
  }

  protected _error(e: Error) {
      this.destination.next({type: 'error', payload: e.message});
      this.destination.complete();
  }

  protected _complete() {
      this.destination.next({type: 'complete'});
      this.destination.complete();
  }
}

Observable.prototype.toDiff = toDiff;
