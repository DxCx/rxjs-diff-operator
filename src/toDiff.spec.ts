'use strict';

import 'jest';
require('babel-core/register');
require('babel-polyfill');

import { Observable } from 'rxjs';
import './toDiff';

describe('toDiff operator', () => {
    it('Should be pass sanity', () => {
        expect(typeof (<any>Observable.prototype).toDiff).toBe('function');
    });

    it('emits basic changes', () => {
        let obs: Observable<number> = Observable.of(1, 2, 3, 4, 5);

        return obs.toDiff().bufferCount(6).toPromise().then(msgs => {
            expect(typeof msgs).toBe('object');
            expect(msgs[0].type).toBe('init');
            expect(msgs[5].type).toBe('complete');
            expect(msgs).toMatchSnapshot();
        });
    });

    it('emits object changes', () => {
        let obs: Observable<{ value: number }> = Observable.of(1, 2, 3, 4, 5).map((v) => ({ value: v}));

        return obs.toDiff().bufferCount(6).toPromise().then(msgs => {
            expect(typeof msgs).toBe('object');
            expect(msgs[0].type).toBe('init');
            expect(msgs[5].type).toBe('complete');
            expect(msgs).toMatchSnapshot();
        });
    });
});
