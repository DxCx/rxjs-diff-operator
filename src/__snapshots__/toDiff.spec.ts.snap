// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toDiff operator emits basic changes 1`] = `
Array [
  Object {
    "isObject": false,
    "payload": 1,
    "type": "init",
  },
  Object {
    "payload": 2,
    "type": "update",
  },
  Object {
    "payload": 3,
    "type": "update",
  },
  Object {
    "payload": 4,
    "type": "update",
  },
  Object {
    "payload": 5,
    "type": "update",
  },
  Object {
    "type": "complete",
  },
]
`;

exports[`toDiff operator emits errors as well 1`] = `
Array [
  Object {
    "isObject": true,
    "payload": Object {
      "value": 1,
    },
    "type": "init",
  },
  Object {
    "payload": Array [
      DiffEdit {
        "kind": "E",
        "lhs": 1,
        "path": Array [
          "value",
        ],
        "rhs": 2,
      },
    ],
    "type": "update",
  },
  Object {
    "payload": "testing errors",
    "type": "error",
  },
]
`;

exports[`toDiff operator emits object changes 1`] = `
Array [
  Object {
    "isObject": true,
    "payload": Object {
      "value": 1,
    },
    "type": "init",
  },
  Object {
    "payload": Array [
      DiffEdit {
        "kind": "E",
        "lhs": 1,
        "path": Array [
          "value",
        ],
        "rhs": 2,
      },
    ],
    "type": "update",
  },
  Object {
    "payload": Array [
      DiffEdit {
        "kind": "E",
        "lhs": 2,
        "path": Array [
          "value",
        ],
        "rhs": 3,
      },
    ],
    "type": "update",
  },
  Object {
    "payload": Array [
      DiffEdit {
        "kind": "E",
        "lhs": 3,
        "path": Array [
          "value",
        ],
        "rhs": 4,
      },
    ],
    "type": "update",
  },
  Object {
    "payload": Array [
      DiffEdit {
        "kind": "E",
        "lhs": 4,
        "path": Array [
          "value",
        ],
        "rhs": 5,
      },
    ],
    "type": "update",
  },
  Object {
    "type": "complete",
  },
]
`;
